aten/src/ATen/core/type_conversion.h

```diff
@@ -0,0 +1,217 @@
+#pragma once
+
+#include <ATen/core/jit_type.h>
+#include <unordered_map>
+#include <atomic>
+
+template <class ValueType>
+class TypeMap {
+  // Internally, we'll use a hash table to store mapping from type
+  // IDs to the values.
+  typedef std::unordered_map<int, ValueType> InternalMap;
+public:
+  typedef typename InternalMap::iterator iterator;
+  typedef typename InternalMap::const_iterator const_iterator;
+  typedef typename InternalMap::value_type value_type;
+
+  const_iterator begin() const { return m_map.begin(); }
+  const_iterator end() const { return m_map.end();  }
+  iterator begin() { return m_map.begin();  }
+  iterator end() { return m_map.end(); }
+
+  // Finds the value associated with the type "Key" in the type map.
+  template <class Key>
+  iterator find() { return m_map.find(getTypeId<Key>());  }
+
+  // Same as above, const version
+  template <class Key>
+  const_iterator find() const { return m_map.find(getTypeId<Key>()); }
+
+  // Associates a value with the type "Key"
+  template <class Key>
+  void put(ValueType &&value) {
+    m_map[getTypeId<Key>()] = std::forward<ValueType>(value);
+  }
+  int size() {
+      return m_map.size();
+  }
+
+private:
+  template <class Key>
+  inline static int getTypeId() {
+    static const int id = LastTypeId++;
+    return id;
+  }
+
+  static std::atomic_int LastTypeId;
+  InternalMap m_map;
+};
+
+template <class ValueType>
+std::atomic_int TypeMap<ValueType>::LastTypeId(0);
```

Comment:
Pretty sure you would end up with one of these atomic ints per type per-compilation unit, which can cause bugs

Reply:
Yeah this was our whole conversation about ODR and stuff like that :/

I think the only solution we know is to use a macro, but I think we decided to go with this for now, and add a macro if we decide we really need to.